stateful/stateful.go

```diff
@@ -0,0 +1,105 @@
+// Package stateful provides stateful HTTP forwarding (i.e. maintaining a 1 to 1
+// relationship between downstream and upstream connections). Clients wishing to
+// take advantage of this capability need to send an initial GET request
+// (analogous to a CONNECT request) that includes the desired host and the HTTP
+// header "X-Lantern-Stateful: true".
+package stateful
+
+import (
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/getlantern/errors"
+	"github.com/getlantern/golog"
+	"github.com/getlantern/idletiming"
+	"github.com/getlantern/interceptor"
+	"github.com/getlantern/ops"
+
+	"github.com/getlantern/http-proxy/filters"
+)
+
+var log = golog.LoggerFor("stateful")
+
+const (
+	// XLanternStateful is the X-Lantern-Stateful header that indicates stateful
+	// connections are OK.
+	XLanternStateful = "X-Lantern-Stateful"
+	xForwardedFor    = "X-Forwarded-For"
+)
+
+type Options struct {
+	Force       bool // set to true to use this without the initial http Request
+	IdleTimeout time.Duration
+	Dialer      func(network, address string) (net.Conn, error)
+	OnRequest   func(req *http.Request)
+	OnResponse  func(resp *http.Response, req *http.Request, responseNumber int) *http.Response
+}
+
+type forwarder struct {
+	*Options
+	ic interceptor.Interceptor
+}
+
+func New(opts *Options) filters.Filter {
+	if opts == nil {
```

Comment:
(Nitpicking) Maybe this shouldn't be accepted in first place, and require a `&Options{}` argument.